wrangler: document the new "redirected" configuration feature (#18757)
cloudflare/cloudflare-docs: src/content/docs/workers/wrangler/configuration.mdx

If you change your routes in the dashboard, Wrangler will override them in the next deploy with the routes you have set in your Wrangler configuration file. To manage routes via the Cloudflare dashboard only, remove any route and routes keys from your Wrangler configuration file. Then add `workers_dev = false` to your Wrangler configuration file. For more information, refer to [Deprecations](/workers/wrangler/deprecations/#other-deprecated-behavior).

Wrangler will not delete your secrets (encrypted environment variables) unless you run `wrangler secret delete <key>`.

> ## Generated Wrangler configuration

Reviewer: Worth adding a link to this from the bundling page?

Author: Cool.

Author: https://github.com/cloudflare/cloudflare-docs/pull/18757/commits/97857ae5393a224e8783d9b210c77d6b23869850

Wrangler will display messaging to the user to indicate that the configuration has been redirected to a different file than the user's configuration file.

### Custom build tool example

A common approach that a build tool might choose to implement is to output code and configuration in a `dist` directory.

- The user writes code that uses Cloudflare Workers resources, configured via a user `wrangler.toml` file.

  ```toml
  name = "my-worker"
  main = "src/index.ts"
  [[kv_namespaces]]
  binding = "<BINDING_NAME1>"
  id = "<NAMESPACE_ID1>"
  ```

  Note that this configuration points `main` at the user code entry-point.

- The user runs a custom build, which might read the user's `wrangler.toml` to find the source code entry-point:

  ```bash
  > my-tool build
  ```

- This `my-tool` generates a `dist` directory that contains both compiled code and a new deployment configuration file.
  It also generates a `.wrangler/deploy/config.json` file that redirects Wrangler to the new generated deployment configuration file:

>   ```plain

Reviewer: What about using https://starlight.astro.build/components/file-tree/?

Author: TIL!

Author: https://github.com/cloudflare/cloudflare-docs/pull/18757/commits/97857ae5393a224e8783d9b210c77d6b23869850

If you change your routes in the dashboard, Wrangler will override them in the next deploy with the routes you have set in your Wrangler configuration file. To manage routes via the Cloudflare dashboard only, remove any route and routes keys from your Wrangler configuration file. Then add `workers_dev = false` to your Wrangler configuration file. For more information, refer to [Deprecations](/workers/wrangler/deprecations/#other-deprecated-behavior).

Wrangler will not delete your secrets (encrypted environment variables) unless you run `wrangler secret delete <key>`.

## Generated Wrangler configuration

:::note

This section describes a feature that can be implemented by frameworks and other build tools that are integrating with Wrangler.

> It is unlikely that an application developer will need to use this feature, but it is documented here to help you understand when Wrangler is using a generated configuration rather than a user's configuration.

Reviewer 2: ```suggestion
It is unlikely that an application developer will need to use this feature, but it is documented here to help you understand when Wrangler is using a generated configuration rather than the original, user's configuration.
```

:::

> When using a framework or a custom pre-build process, some tools need to generate a modified Wrangler configuration alongside the generated code.

Reviewer 2: ```suggestion
When using a framework or a custom pre-build process, some tools need to generate a new Wrangler configuration alongside the original code.
```

Author: In this case we actually want to say alongside generated code. The "generated config" may get stored on disk in the same directory as the "generated/compiled source code".

Author: I attempted to improve that.

> In this case, the tool may also create a special `.wrangler/deploy/config.json` file that redirects Wrangler to use the generated configuration rather than the original user configuration.

Reviewer 2: ```suggestion
In this case, the tool may also create a special `.wrangler/deploy/config.json` file that redirects Wrangler to use the generated configuration rather than the original, user's configuration.
```

Wrangler uses this generated configuration only for the following deploy and dev related commands:

- `wrangler deploy`
- `wrangler dev`
- `wrangler versions upload`
- `wrangler versions deploy`
- `wrangler pages deploy`
- `wrangler pages build`
- `wrangler pages build-env`

When running these commands, Wrangler looks up the directory tree from the current working directory for a file at the path `.wrangler/deploy/config.json`.
This file must contain only a single JSON object of the form:

```json
{ "configPath": "../../path/to/wrangler.json" }
```

> When this `config.json` file exists Wrangler will follow the `configPath` (relative to the `.wrangler/deploy/config.json` file) to find the generated Wrangler configuration file to load and use in the current command.

Reviewer 2: ```suggestion
When this `config.json` file exists, Wrangler will follow the `configPath` (relative to the `.wrangler/deploy/config.json` file) to find the generated Wrangler configuration file to load and use in the current command.
```

Wrangler will display messaging to the user to indicate that the configuration has been redirected to a different file than the user's configuration file.

### Custom build tool example

> A common approach that a build tool might choose to implement is to output code and configuration in a `dist` directory.

Reviewer 2: ```suggestion
For example, a new configuration file may be generated by a custom build tool when creating a `dist` directory.
```
You may need to rewrite this if it doesn't work. I added this because I didn't quite understand the connection between this and the section above. I know it says "example" in the header, but I think it would also help to lead in by saying specifically saying that this is an example of when a new configuration file is generated.

Author: Updated to make it more example-ish

> - The user writes code that uses Cloudflare Workers resources, configured via a user `wrangler.toml` file.

Reviewer 2: ```suggestion
- First, the user writes code that uses Cloudflare Workers resources, configured via a user `wrangler.toml` file.
```
I'm just adding "first" and "then" to make it more clear that this is chronological, rather than a list. You can also consider numbered steps, but I'm not sure that makes sense in this case.

  ```toml
  name = "my-worker"
  main = "src/index.ts"
  [[kv_namespaces]]
  binding = "<BINDING_NAME1>"
  id = "<NAMESPACE_ID1>"
  ```

  Note that this configuration points `main` at the user code entry-point.

> - The user runs a custom build, which might read the user's `wrangler.toml` to find the source code entry-point:

Reviewer 2: ```suggestion
- Then, the user runs a custom build, which might read the user's `wrangler.toml` to find the source code entry-point:
```

  ```bash
  > my-tool build
  ```

- This `my-tool` generates a `dist` directory that contains both compiled code and a new deployment configuration file.

>   It also generates a `.wrangler/deploy/config.json` file that redirects Wrangler to the new generated deployment configuration file:

Reviewer 2: ```suggestion
  It also generates a `.wrangler/deploy/config.json` file that redirects Wrangler to the new, generated deployment configuration file:
```
